hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java

```diff
@@ -573,54 +624,85 @@ public HoodieInstant transitionReplaceRequestedToInflight(HoodieInstant requeste
     // Then write to timeline
-    transitionState(requestedInstant, inflightInstant, data);
+    transitionPendingState(requestedInstant, inflightInstant, data);
     return inflightInstant;
   }
 
   /**
    * Transition replace inflight to Committed.
    *
+   * @param shouldLock Whether to hold the lock when performing transition
    * @param inflightInstant Inflight instant
    * @param data Extra Metadata
    * @return commit instant
    */
-  public HoodieInstant transitionReplaceInflightToComplete(HoodieInstant inflightInstant, Option<byte[]> data) {
+  public HoodieInstant transitionReplaceInflightToComplete(boolean shouldLock,
+                                                           HoodieInstant inflightInstant, Option<byte[]> data) {
     ValidationUtils.checkArgument(inflightInstant.getAction().equals(HoodieTimeline.REPLACE_COMMIT_ACTION));
     ValidationUtils.checkArgument(inflightInstant.isInflight());
     HoodieInstant commitInstant = new HoodieInstant(State.COMPLETED, REPLACE_COMMIT_ACTION, inflightInstant.getTimestamp());
     // Then write to timeline
-    transitionState(inflightInstant, commitInstant, data);
+    transitionStateToComplete(shouldLock, inflightInstant, commitInstant, data);
     return commitInstant;
   }
 
-  private void transitionState(HoodieInstant fromInstant, HoodieInstant toInstant, Option<byte[]> data) {
-    transitionState(fromInstant, toInstant, data, false);
+  private void transitionPendingState(HoodieInstant fromInstant, HoodieInstant toInstant, Option<byte[]> data) {
+    transitionPendingState(fromInstant, toInstant, data, false);
+  }
+
+  protected void transitionStateToComplete(boolean shouldLock, HoodieInstant fromInstant,
+                                           HoodieInstant toInstant, Option<byte[]> data) {
+    ValidationUtils.checkArgument(fromInstant.getTimestamp().equals(toInstant.getTimestamp()), String.format("%s and %s are not consistent when transition state.", fromInstant, toInstant));
+    String fromInstantFileName = fromInstant.getFileName();
+    // Ensures old state exists in timeline
+    LOG.info("Checking for file exists ?" + getInstantFileNamePath(fromInstantFileName));
+    try {
+      if (metaClient.getTimelineLayoutVersion().isNullVersion()) {
+        // Re-create the .inflight file by opening a new file and write the commit metadata in
+        createFileInMetaPath(fromInstantFileName, data, false);
+        Path fromInstantPath = getInstantFileNamePath(fromInstantFileName);
+        HoodieInstant instantWithCompletionTime = new HoodieInstant(toInstant.getState(), toInstant.getAction(),
+            toInstant.getTimestamp(), metaClient.createNewInstantTime(false));
+        Path toInstantPath = getInstantFileNamePath(instantWithCompletionTime.getFileName());
+        boolean success = metaClient.getFs().rename(fromInstantPath, toInstantPath);
```

Comment:
we should be avoiding renames, for cloud storage? I think we should make higher level APIs like 

HoodieStorage.writeFileAtomic(..)  and implement differently for hdfs and cloud storage say.